### ISSUE TITLE
Create Audio Sink in Unity

### ISSUE THREAD

**Issue by sebjf** [pixiv/webrtc#14]
Hi, 

I would like to receive audio from the browser in Unity. I see it is possible to add a sink to a track passed to OnTrack using the AudioTrackInterfaceExtension::AddSink() method. However, at present there is no class implementing IAudioTrackSinkInterface that allows native code to callback a managed OnData method?

So to receive audio I must:

1. Create a native class DelegatingAudioSinkInterface that subclasses webrtc::AudioTrackSinkInterface.
2. Have the subclass override OnData and callback into a managed object using a function pointer passed in on construction (as the DelegatingVideoSyncInterface does).
3. Create a managed DisposableAudioSinkInterface class, that constructs DelegatingAudioSinkInterface using a C api. 
4. DisposableAudioSinkInterface should receive a IManagedAudioSink object, and retrieve the function pointers from this to pass via the C api to the new DelegatingAudioSinkInterface. 
5. The C api should return a C pointer to DelegatingAudioSinkInterface, which stores it and returns it through the Ptr member, implementing IAudioTrackSinkInterface. 
6. The DisposableAudioSinkInterface can then be passed to the AudioTrackInterfaceExtension AddSink method. 

There is already a managed type AudioTrackSinkInterface, that is created by the AddSink callback from a native pointer passed from DelegatingAudioSourceInterface. This should not be used because that sink is owned by WebRTC. Instead a new disposable class should be created to manage the lifetime of the managed sink.

Is this correct?

**Comment by akihikodaki** [comment]
I'm sorry to tell you that would not work well.

tl; dr: The audio processing is required to be in real time and the garbage-collector Unity depends on is not really good at such a workload.

First, you have to understand how threads and the garbage collector interact. Each threads managed by Mono runtime holds references to allocated objects, and the garbage collector marks those references so that it does not free the objects still referenced.  (Note: you still depend on Mono runtime even if you use IL2CPP; IL2CPP only replaces the IL evaluation feature of the runtime.) The mark process requires the threads to be stopped (i.e., to "stop the world"). See the following article to learn how the behavior can be observed (and the incremental garbage collection mitigates the impact):
https://blogs.unity3d.com/2018/11/26/feature-preview-incremental-garbage-collection/

Now, let's get back to WebRTC. `webrtc::AudioTrackSinkInterface.OnData` is called from a special thread with a high priority to minimize audio disruptions because humans are sensitive for them. One problem is the thread is not managed by Mono runtime. Fortunately, Mono provides a solution for this case: call `mono_thread_attach()` to let Mono manage the thread.
https://www.mono-project.com/docs/advanced/embedding/#threading-issues
However, this also causes other two problems:
- `mono_thread_attach()` is the function Unity manages and hard to refer to it for an external library.
- The thread will be affected by "stop-the-world", which is not good for real-time processing.

The later problems is critical and the very reason I decided not to implement `IManagedAudioSink`.

It still does not mean you have to abandon audio completely. There are possible alternative solutions:
- Do it completely in C++ if you do not need Unity for audio processing.
- Implement a native audio plugin, an infrastructure Unity provides for real-time audio, if you need to incorporate WebRTC with Unity's audio. See https://docs.unity3d.com/Manual/AudioMixerNativeAudioPlugin.html.

The later option introduces a dependency on Unity. Either way, the implementation will be more application-dependent and not be suitable to live in this repository.

**Comment by sebjf** [comment]
Hi, Thank you for explaining. 

I believe I understand: OnData callbacks are made from a thread unknown by mono, meaning they may try to access variables being moved by the GC. To prevent this, mono must be given control of the thread, but this is difficult and introduces side-effects.
(The problem does not occur for video because OnFrame is called by one of the threads that are wrapped?)

Could another thing to try then be to receive OnData though a delegate [1] that limits access to pinned objects? This could then attempt to marshal the data into OnFilterRead().

Obviously that is out of scope of this repository.

However, what about just the C API?

If I,

1. Create a DelegatingAudioSinkInterface class that stores a function pointer for the OnData callback,
2. Create a C API that creates/destroys this object (create call taking only a function pointer retrieved with GetFunctionPointerForDelegate)

This should follow the existing binding conventions?

However, no DisposableAudioSinkInterface or IManagedAudioSink would be created. Instead, the pointer returned by (2) could be passed to an instance of AudioTrackSinkInterface, which in turn could be provided to the IAudioTrackInterface.AddSink() extension. 

This would be done in application code through PInvoke.

A native plugin could also find the address of the function to interact with it, without needing to link explicitly.

Would that work?


[1] Through a delegate that is kept around longer than the sink, or a static method.

**Comment by akihikodaki** [comment]
> I believe I understand: OnData callbacks are made from a thread unknown by mono, meaning they may try to access variables being moved by the GC. To prevent this, mono must be given control of the thread, but this is difficult and introduces side-effects.

Yes, that's right.

>  (The problem does not occur for video because OnFrame is called by one of the threads that are wrapped?)

Actually, `OnFrame` has the same problem. A C++ class, `webrtc::VideoBuffer` and its .NET binding, `Pixiv.Webrtc.VideoBuffer` are added to workaround this. The example code which resides at `examples/unity` polls the pointer `webrtc::VideoBuffer` holds to maintain the reference to the latest frame provided by `OnFrame`. This works because skipping a few frames is often fine for video.

> Could another thing to try then be to receive OnData though a delegate [1] that limits access to pinned objects? This could then attempt to marshal the data into OnFilterRead().

I cannot be sure about that. Mono creates a bunch of metadata of a thread when `mono_thread_attach` is called and there is no clear way to ensure it does not refer to them when you do not call `mono_thread_attach`. I'm also not sure if static fields are really "static"; maybe the underlying memory region is managed by the GC. (Also remember: Unity can break assumptions you have. In fact, I was forced to dig into https://github.com/Unity-Technologies/mono a few times when I was developing a Unity app.)

> However, what about just the C API?
> 
> If I,
> 
>   1. Create a DelegatingAudioSinkInterface class that stores a function pointer for the OnData callback,
> 
>   2. Create a C API that creates/destroys this object (create call taking only a function pointer retrieved with GetFunctionPointerForDelegate)
> 
> 
> This should follow the existing binding conventions?
> 
> However, no DisposableAudioSinkInterface or IManagedAudioSink would be created. Instead, the pointer returned by (2) could be passed to an instance of AudioTrackSinkInterface, which in turn could be provided to the IAudioTrackInterface.AddSink() extension.
> 
> This would be done in application code through PInvoke.
> 
> A native plugin could also find the address of the function to interact with it, without needing to link explicitly.
> 
> Would that work?

I belive that would work. However, I recommend you just to link this library or to modify it. Linking the library has an obvious benefit that it can make sure the function actually exist at the build time. Modifying this library is completely fine, too; I'm somewhat certain that reworking a patch to make it consistent with WebRTC's update is easier than maintaining your own build infrastructure.
